add node relationships

moolitayer opened the pull request.

moolitayer commented:
fyi @abonas @simon3z
do we want to add node count in the project explorer view as well?

I will check if distinct is needed tomorrow. hence wip
@miq-bot add_label wip

![screenshot from 2015-09-20 18-30-59](https://cloud.githubusercontent.com/assets/3010449/9981645/0edd2dca-5fcb-11e5-809e-75faef672c96.png)
![screenshot from 2015-09-20 18-48-19](https://cloud.githubusercontent.com/assets/3010449/9981644/0edbed8e-5fcb-11e5-8028-69be18f25576.png)
![screenshot from 2015-09-20 18-57-17](https://cloud.githubusercontent.com/assets/3010449/9981646/0ede9368-5fcb-11e5-8a66-99ff08f2e756.png)
![screenshot from 2015-09-20 19-00-59](https://cloud.githubusercontent.com/assets/3010449/9981647/0edee980-5fcb-11e5-8c22-8e567e8e5dd2.png)

abonas commented:
@miq_bot add_label providers/containers

abonas commented:
@moolitayer I suggest providing a more meaningful description/commit message "This patch adds the backend impl for relationship between X to Y for A, B C, as well as its UI representation"

abonas commented:
@miq-bot add_label ui

moolitayer commented:
> @moolitayer I suggest providing a more meaningful description/commit message "This patch adds the backend impl for relationship between X to Y for A, B C, as well as its UI representation"
> And how about adding some tests?

What is the right place/scope for such tests?
do you mean for the ui/relationship?

As for the commit message, I've written meaningful commit messages in the four commit in this pr:
* project item view: add node relationship
* route item view: add node relationship
* node item view: add route relationship
* image item view: add node relationship
I don't like to copy the different items into the pr, especially since then when changes happen you have to change multiple places and some might be forgetten

moolitayer commented:
address rubocop's comment.

abonas commented:
> What is the right place/scope for such tests?
> do you mean for the ui/relationship?

for the relationship backend part at minimum.

> As for the commit message, I've written meaningful commit messages in the four commit in this pr:
why is there a need in 4 commits in this PR? 
> project item view: add node relationship
> route item view: add node relationship
> node item view: add route relationship
> image item view: add node relationship 
> I don't like to copy the different items into the pr, especially since then when changes happen you have to change multiple places and some might be forgetten

writing in the commit message "view" suggests this is UI related change only, not backend as well.

moolitayer commented:
updated commit messages, added tests

> why is there a need in 4 commits in this PR?

This Is my understanding of how commits should work. There is no real connection between:
```project item view: add node relationship``` to  ```node item view: add route relationship``` they are just "more of the same" and that's why they are in the same pull request

moolitayer commented:
@abonas are you ok with this change?

abonas commented:
I think that "distinct" should be added for at least some of the relationships in this PR. Please check.

moolitayer commented:
> I think that "distinct" should be added for at least some of the relationships in this PR. Please check.

right, updated. I left it out in:
node -> service -> route: since a route is defined by the service it belongs to
route -> service -> nodes: since a route belongs to one service

moolitayer commented:
@miq-bot remove_label wip

chessbyte commented:
@moolitayer Please rebase and push again

moolitayer commented:
rebased 
fyi @abonas @simon3z

blomquisg commented:
@moolitayer is there anything needed for the inventory refresh to ensure that these relationships are populated?

moolitayer commented:
@blomquisg no. all relay on has_many - through of existing entities.

simon3z commented on app/helpers/container_node_helper/textual_summary.rb:
```diff
@@ -10,7 +10,7 @@ def textual_group_properties
 
   def textual_group_relationships
-    %i(ems container_services container_replicators container_groups containers lives_on)
+    %i(ems container_services container_routes container_replicators container_groups containers lives_on)
```
Order should be: `%i(ems container_routes container_services ...`

simon3z commented on app/models/container_image.rb:
```diff
@@ -7,6 +7,7 @@ class ContainerImage < ActiveRecord::Base
   belongs_to :container_image_registry
   belongs_to :ext_management_system, :foreign_key => "ems_id"
   has_many :containers
+  has_many :container_nodes, -> { distinct }, :through => :containers
```
@moolitayer please add a test for this.

Use `FactoryGirl` to create two (or more) containers with the same image and located on the same node. Then check that the `container_nodes` has no duplicated entries.

simon3z commented on app/models/container_node.rb:
```diff
@@ -8,6 +8,7 @@ class ContainerNode < ActiveRecord::Base
   has_many   :container_conditions, :class_name => ContainerCondition, :as => :container_entity, :dependent => :destroy
   has_many   :containers, :through => :container_groups
   has_many   :container_services, -> { distinct }, :through => :container_groups
+  has_many   :container_routes, :through => :container_services
```
@moolitayer add a test here as well (demonstrating that there's no need for `distinct`).

simon3z commented on app/models/container_project.rb:
```diff
@@ -7,6 +7,7 @@ class ContainerProject < ActiveRecord::Base
   has_many :container_replicators
   has_many :container_services
   has_many :container_definitions, :through => :container_groups
+  has_many :container_nodes, -> { distinct }, :through => :container_groups
```
@moolitayer add a test.

simon3z commented on app/models/container_route.rb:
```diff
@@ -5,6 +5,8 @@ class ContainerRoute < ActiveRecord::Base
   belongs_to :ext_management_system, :foreign_key => "ems_id"
   belongs_to :container_project
   belongs_to :container_service
+  has_many :container_nodes, :through => :container_service
```
@moolitayer add a test

simon3z commented:
@moolitayer I think there's a valid complain from travis to solve.

moolitayer commented:
fix failing test. reorder container_node_helper/textual_summary.rb.
@simon3z  add tests for image => node and project => node relatioships

I don't see how I can test the node => route and route => node relationships. If I will add such a test it will fail; I'm relaying on the fact that openshift always defines a new route per each service.

please review

simon3z commented:
> I don't see how I can test the node => route and route => node relationships. If I will add such a test it will fail; I'm relaying on the fact that openshift always defines a new route per each service.

@moolitayer will `distict` solve this case as well?

moolitayer commented:
added tests. now we have tests for all 4 relationships.

Note: to my surprise node -> route needed a distinct definition as well.
although the relation is defined this way:
route -> service -> groups -> nodes
and 
service -> groups -> nodes is already defined as distinct.
the same applies to route -> node as well(same relation the other way).
@simon3z please review

chessbyte commented on spec/models/manageiq/providers/kubernetes/container_manager/entity_spec.rb:
```diff
@@ -0,0 +1,61 @@
+require "spec_helper"
+
+describe ContainerProject do
```
why are all these tests for various classes in a single spec file, instead of being split out into a spec per class/model?

moolitayer commented:
Tests have been split to individual files.

simon3z commented:
@moolitayer, just to make sure, so `distinct` was required in all cases? (without it tests are failing?)

moolitayer commented:
> @moolitayer, just to make sure, so distinct was required in all cases? (without it tests are failing?)

@simon3z yes, it surprised me. wrote about it in https://github.com/ManageIQ/manageiq/pull/4446#issuecomment-147751279

simon3z commented:
LGTM :+1: cc @chessbyte ready for merge imo
